pkg/kclient/all.go

```diff
@@ -0,0 +1,182 @@
+	"k8s.io/klog"
+)
+
+// Code into this file is heavily inspired from https://github.com/ahmetb/kubectl-tree
+
+func (c *Client) GetAllResourcesFromSelector(selector string, ns string) ([]unstructured.Unstructured, error) {
+	apis, err := findAPIs(c.discoveryClient)
+	if err != nil {
+		return nil, err
+	}
+	return getAllResources(c.DynamicClient, apis.list, ns, selector)
+}
+
+func getAllResources(client dynamic.Interface, apis []apiResource, ns string, selector string) ([]unstructured.Unstructured, error) {
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	var out []unstructured.Unstructured
+
+	start := time.Now()
+	klog.V(2).Infof("starting to query %d APIs in concurrently", len(apis))
+
+	var errResult error
+	for _, api := range apis {
+		if !api.r.Namespaced {
+			klog.V(4).Infof("[query api] api (%s) is non-namespaced, skipping", api.r.Name)
+			continue
+		}
+		wg.Add(1)
+		go func(a apiResource) {
+			defer wg.Done()
+			klog.V(4).Infof("[query api] start: %s", a.GroupVersionResource())
+			v, err := queryAPI(client, a, ns, selector)
+			if err != nil {
+				klog.V(4).Infof("[query api] error querying: %s, error=%v", a.GroupVersionResource(), err)
+				errResult = err
+				return
+			}
+			mu.Lock()
+			out = append(out, v...)
+			mu.Unlock()
+			klog.V(4).Infof("[query api]  done: %s, found %d apis", a.GroupVersionResource(), len(v))
+		}(api)
+	}
+
+	klog.V(2).Infof("fired up all goroutines to query APIs")
+	wg.Wait()
+	klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
+	klog.V(2).Infof("query result: error=%v, objects=%d", errResult, len(out))
+	return out, errResult
+}
+
+func queryAPI(client dynamic.Interface, api apiResource, ns string, selector string) ([]unstructured.Unstructured, error) {
+	var out []unstructured.Unstructured
+
+	var next string
+	for {
+		var intf dynamic.ResourceInterface
+		nintf := client.Resource(api.GroupVersionResource())
+		intf = nintf.Namespace(ns)
+		resp, err := intf.List(context.TODO(), metav1.ListOptions{
+			Limit:         250,
+			Continue:      next,
+			LabelSelector: selector,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("listing resources failed (%s): %w", api.GroupVersionResource(), err)
+		}
+		out = append(out, resp.Items...)
+
+		next = resp.GetContinue()
+		if next == "" {
+			break
+		}
+	}
+	return out, nil
+}
+
+type apiResource struct {
+	r  metav1.APIResource
+	gv schema.GroupVersion
+}
+
+func (a apiResource) GroupVersionResource() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    a.gv.Group,
+		Version:  a.gv.Version,
+		Resource: a.r.Name,
+	}
+}
+
+type resourceNameLookup map[string][]apiResource
+
+type resourceMap struct {
+	list []apiResource
+	m    resourceNameLookup
+}
+
+func findAPIs(client discovery.DiscoveryInterface) (*resourceMap, error) {
+	start := time.Now()
+	resList, err := client.ServerPreferredResources()
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch api groups from kubernetes: %w", err)
```

Comment:
IMO, we should return a zero value instead of `nil` when returning a pointer because the program would panic if the caller were to make use of it. I know it's not the case here, and that most of the time if `err != nil`, we don't care about what else was returned. I'm saying it so that we don't get into a problem if the returned value is indeed used.

Correct me if I am wrong.

Reply:
IMO, absolutely not. The accepted rule is that if `err` is not nil, the other values are not usable. 
If the caller tries to use this value, it is a programming error and a potential bug/security hole, and should by caught as soon as possible. The panic is a very good way to catch it.

Reply:
Hmm. Makes sense. 👍🏾